src/billing/request_payment.ts

```diff
@@ -1,78 +1,102 @@
 
-export async function requestPayment(
-  session: SessionInterface,
-  isTest: boolean,
-): Promise<string | undefined> {
-  const returnUrl = buildEmbeddedAppUrl(
-    Buffer.from(`${session.shop}/admin`).toString('base64'),
-  );
-
-  let data: RequestResponse;
-  if (isRecurring()) {
-    const mutationResponse = await requestRecurringPayment(
-      session,
-      returnUrl,
-      isTest,
-    );
-    data = mutationResponse.data.appSubscriptionCreate;
-  } else {
-    const mutationResponse = await requestSinglePayment(
-      session,
-      returnUrl,
-      isTest,
+interface RequestPaymentParams {
+  session: SessionInterface;
+  isTest: boolean;
+}
+
+interface RequestPaymentInternalParams {
+  config: BillingConfig;
+  client: InstanceType<ReturnType<typeof createGraphqlClientClass>>;
+  returnUrl: string;
+  isTest: boolean;
+}
+
+export function createRequestPayment(config: ConfigInterface) {
+  return async function ({
+    session,
+    isTest,
+  }: RequestPaymentParams): Promise<string | undefined> {
+    if (!config.billing) {
+      throw new BillingError({
+        message: 'Attempted to request payment without billing configs',
+        errorData: [],
+      });
+    }
+
+    const returnUrl = createBuildEmbeddedAppUrl(config)(
+      hashString(`${session.shop}/admin`, HashFormat.Base64),
     );
-    data = mutationResponse.data.appPurchaseOneTimeCreate;
-  }
 
-  if (data.userErrors?.length) {
-    throw new BillingError({
-      message: 'Error while billing the store',
-      errorData: data.userErrors,
+    const GraphqlClient = createGraphqlClientClass({config});
+    const client = new GraphqlClient({
+      domain: session.shop,
+      accessToken: session.accessToken,
     });
-  }
 
-  return data.confirmationUrl;
-}
+    let data: RequestResponse;
+    if (isRecurring(config.billing)) {
+      const mutationResponse = await requestRecurringPayment({
+        config: config.billing,
+        client,
+        returnUrl,
+        isTest,
+      });
+      data = mutationResponse.data.appSubscriptionCreate;
+    } else {
+      const mutationResponse = await requestSinglePayment({
+        config: config.billing,
+        client,
+        returnUrl,
+        isTest,
+      });
+      data = mutationResponse.data.appPurchaseOneTimeCreate;
+    }
 
-async function requestRecurringPayment(
-  session: SessionInterface,
-  returnUrl: string,
-  isTest: boolean,
-): Promise<RecurringPaymentResponse> {
-  if (!config.billing) {
-    throw new BillingError({
-      message: 'Attempted to request recurring payment without billing configs',
-      errorData: [],
-    });
-  }
+    if (data.userErrors?.length) {
+      throw new BillingError({
+        message: 'Error while billing the store',
+        errorData: data.userErrors,
+      });
+    }
 
-  const client = new GraphqlClient(session.shop, session.accessToken);
+    return data.confirmationUrl;
+  };
+}
 
+async function requestRecurringPayment({
+  config,
```

Comment:
Again, this `config` refers to `BillingConfig`, not a `ConfigInterface` config, which is kinda confusing given that both are referenced in the one file.